Allow to deprecate options and aliases by date

felixfontein opened the pull request:
##### SUMMARY
Allows to deprecate options and aliases also by date and not only by version.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/basic.py
lib/ansible/module_utils/common/parameters.py
lib/ansible/module_utils/common/warnings.py
lib/ansible/module_utils/csharp/Ansible.Basic.cs
test/lib/ansible_test/_data/sanity/pylint/plugins/deprecated.py
test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py
test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py

felixfontein committed:
Allow to deprecate options and aliases by date instead of only by version.

felixfontein committed:
Update display.deprecate().

felixfontein committed:
Adjust behavior to conform to tested behavior, extend tests, and improve C# style.

felixfontein committed:
Parse date and fail on invalid date.

This is mainly to make sure that people start using invalid dates, and we eventually have a mess to clean up.

felixfontein committed:
C# code: improve validation and update/extend tests.

felixfontein committed:
Make sure that deprecate() is not called with both date and version.

felixfontein committed:
Forgot to remove no longer necessary formatting.

felixfontein committed:
Adjust order of warnings in C# code.

felixfontein committed:
Adjust unrelated test.

felixfontein committed:
Fix grammar (and make that test pass).

felixfontein commented:
(The remaining CI failures are package download timeouts, i.e. unrelated :) )

ready_for_review

felixfontein commented:
TODO next: make more compatible with #67684, don't validate at runtime, validate dates in sanity checks

felixfontein committed:
Don't parse date, and adjust message to be same as in #67684.

felixfontein committed:
Sanity tests: disable date in past test.

felixfontein committed:
Validate-modules: validate ISO 8601 date format.

felixfontein committed:
Validate-modules: switch schema declaration for deprecated_aliases to improve error messages for invalid dates.

mattclay commented:
@felixfontein What's the intended workflow for this feature? Deprecation by version seems obvious to me, but I'm not immediately coming up with a scenario for deprecation by date.

felixfontein commented:
@mattclay I heared some time ago from @sivel (IIRC) that the idea was that especially in collections, it is easier to deprecate by date then by version, since it is easier to understand for users when something is actually removed. Especially since there no longer is the "regular" 6 month Ansible release cycle. (Also people might have no clue what "removed in version x.yz" means when they don't really know in which collection the module is, and how that collection does versioning - I guess that will be common for modules from stable-2.9 that users will continue to use with the ACD, which allows users to ignore the whole collection business.)

bcoca commented on lib/ansible/module_utils/basic.py:
```diff
@@ -720,9 +720,16 @@ def warn(self, warning):
+        if date is not None:
+            self.log('[DEPRECATION WARNING] %s %s' % (msg, date))
+        else:
+            self.log('[DEPRECATION WARNING] %s %s' % (msg, version))
```
instead of having 2 params allow version to be overloaded ?

felixfontein replied:
I would really prefer not to do that. Having two separate parameters makes it easier to validate, since it is clear from the parameter name what the contents must look like.

bcoca replied:
i would at least make them mutually exclusive, really don't want people complaining about precedence of date ><>= version

bcoca replied:
the main reason for using as overload is that we don't backport features, and this clearly is a feature.

sivel replied:
I agree we should keep 2 args so that we can validate in CI that they are formatted correctly. Making the mutually exclusive also sounds like a good plan.

felixfontein replied:
Indeed! And: lines 724 and 725 above raise `AssertionError("implementation error -- version and date must not both be set")` if both are set :)

sivel replied:
haha, was that already there? I didn't look, I just commented on a comment :)

felixfontein replied:
Hehe :) Github's excerpt is unfortunately hiding all the interesting parts and comments explaining these few lines ;)

sivel commented:
> @mattclay I heared some time ago from @sivel (IIRC) that the idea was that especially in collections, it is easier to deprecate by date then by version, since it is easier to understand for users when something is actually removed

This is correct, and this mirrors what is being done in https://github.com/ansible/ansible/pull/67684

Collections will deprecate everything by date.

felixfontein commented:
> Collections will deprecate everything by date.

Is this a guideline, an assumption by ansible-core, or something we want to enforce? Right now (and also with both this PR and #66920), collections can also deprecate by (their own) version. This isn't so great for end-users, but there's no technical reason they should not be able to do that.

sivel commented:
> Is this a guideline, an assumption by ansible-core, or something we want to enforce

@felixfontein in the end, we cannot enforce this for everyone. It's more of a general recommendation or guideline. Any random collection can do whatever they want. However this is a rule that we will follow for collections that we are responsible for in some way.

felixfontein commented:
@sivel sounds good. I mainly wanted to make sure that we don't actively want to sabotage people trying to deprecate by collection version, i.e. don't add some of the collection changes from #66920 :)

mattclay commented:
@felixfontein Before we go farther on the implementation here (specifically for the sanity tests) I'd like to finish working through https://github.com/ansible/ansible/pull/66920 since they're so closely related.

jborean93 commented on lib/ansible/module_utils/csharp/Ansible.Basic.cs:
```diff
@@ -242,10 +243,18 @@ public void Debug(string message)
         }
 
-        public void Deprecate(string message, string version)
+        public void Deprecate(string message, string version = null, string date = null)
```
Unfortunately PowerShell does not handle kwargs in C# methods very well, we cannot do `$module.Deprecate('msg', date='date')` we would have to do `$module.Deprecate('msg', $null, 'date')` which isn't ideal. An alternative is to create an overloaded method

```
public void Deprecate(string message, string version)
{
    Existing logic
}

public void Deprecate(string message, DateTime date)
{
    New logic
}
```

That way a module can differentiate between the 2 with `$module.Deprecate('msg', [DateTime]'2020-03-27')`  or use whatever other date format it wants and the C# method can format that date into a common output format, preferably ISO 8601 `yyyy-mm-dd`.

felixfontein replied:
How do I convert a `DateTime` object to ISO 8601? (Sorry for asking, I had to write that code essentially blind since I have neither C# nor PowerShell available for testing...)

jborean93 replied:
No need to apologise for asking :) A DateTime object's [ToString](https://docs.microsoft.com/en-us/dotnet/api/system.datetime.tostring?view=netframework-4.8#System_DateTime_ToString_System_String_) method can take in a string that provides how to format the date as a string. In this case you want to do `date.ToString("yyyy-MM-dd")`.

The format can be either a known [standard format](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings?view=netframework-4.8) or a [custom format](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings?view=netframework-4.8).

jborean93 commented on lib/ansible/module_utils/csharp/Ansible.Basic.cs:
```diff
@@ -83,6 +83,7 @@ public class AnsibleModule
             { "no_log", new List<object>() { false, typeof(bool) } },
             { "options", new List<object>() { typeof(Hashtable), typeof(Hashtable) } },
             { "removed_in_version", new List<object>() { null, typeof(string) } },
+            { "removed_at_date", new List<object>() { null, typeof(string) } },
```
Would be better as `typeof(DateTime)`. The only downside here is that a user must use the `[DateTime]` cast in PowerShell to ensure the type is DateTime, otherwise it will probably fail when the C# code goes to parse it.

felixfontein replied:
@jborean93 what happens in the validate-modules test, when a argument spec containing `DateTime` objects is exported? Will they be converted properly?

jborean93 replied:
On PowerShell Core (which we use for validate-modules) it will be exported as a structured object that contains the ISO 8601 string and other values

```
PS /home/jborean> @{a = (Get-Date) } | ConvertTo-Json
{
  "a": {
    "value": "2020-03-27T07:02:17.147412+10:00",
    "DisplayHint": 2,
    "DateTime": "Friday, 27 March 2020 7:02:17 am"
  }
}
```

But even if we wanted to follow a similar structure to python or were unhappy with how the defaults look we can add our own custom formatting for a DateTime type here https://github.com/ansible/ansible/blob/devel/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/ps_argspec.ps1#L35.

felixfontein replied:
Does the following look correct?
```
        } elseif ($value -is [DateTime]) {
            $Hash[$key] = $value.ToString("yyyy-MM-dd")
        }
```

jborean93 replied:
yep if you want it to be a string in that format that's how it would look.

felixfontein replied:
Yep. The validation should work the same for both Python and PowerShell modules.

One point though: if a different type than `DateTime` is used, `validate-module` will not detect that correctly. In particular, if a user uses `string` instead of `DateTime`, `validate-module` cannot detect that.

jborean93 replied:
That may be the case but IIRC the AnsibleModule parsing will fail to cast that type to a DateTime so the module wouldn't actually work.

felixfontein replied:
Except for `deprecatedAliases`, since there no check is done. I've added one in ed1da9be147048bf752e9900fe2bd515b687fbf1 (and hope it is correct).

felixfontein replied:
Ok, at least the tests do now pass again :)

jborean93 reviewed (changes requested):
Thanks for actually looking at the Windows stuff and adding tests. Just a few suggestions.

felixfontein committed:
Use DateTime instead of string for date deprecation.

felixfontein committed:
Validate that date in deprecated_aliases is actually a DateTime.

felixfontein committed:
Fix tests.

felixfontein committed:
Fix rebasing error.

felixfontein committed:
Adjust error codes for pylint, and add removed_at_date and deprecated_aliases.date checks to validate-modules.

felixfontein committed:
Make deprecation date in the past error codes optional.

felixfontein committed:
Make sure not both version and date are specified for AnsibleModule.deprecate() calls.

felixfontein committed:
Stop using Python 3.7+ API.

felixfontein committed:
Make sure errors are actually reported. Re-add 'ansible-' prefix.

felixfontein committed:
Avoid crashing when 'name' isn't there.

felixfontein commented:
I've updated the PR, extended/improved/fixed the tests, and changed the behavior so that deprecation by dates in the past are only reported as errors when `--enable-optional-errors` is specified.

ready_for_review

felixfontein committed:
Linting.

felixfontein commented:
/rebuild_failed

jborean93 commented on lib/ansible/module_utils/csharp/Ansible.Basic.cs:
```diff
@@ -689,21 +697,44 @@ private void ValidateArgumentSpec(IDictionary argumentSpec)
                     }
+                    if (!depInfo.ContainsKey("version") && !depInfo.ContainsKey("date"))
+                    {
+                        string msg = "One of version and date is required in a deprecated_aliases entry";
```
```suggestion
                        string msg = "One of version or date is required in a deprecated_aliases entry";
```

jborean93 commented on lib/ansible/utils/display.py:
```diff
@@ -258,6 +259,8 @@ def deprecated(self, msg, version=None, removed=False):
             if version:
                 new_msg = "[DEPRECATION WARNING]: %s. This feature will be removed in version %s." % (msg, version)
+            elif date:
+                new_msg = "[DEPRECATION WARNING]: %s. This feature will be removed in a release after %s." % (msg, date)
```
Should we state that it will be removed in a major release after x like?

```suggestion
                new_msg = "[DEPRECATION WARNING]: %s. This feature will be removed in a major release after %s." % (msg, date)
```

felixfontein replied:
That's true for collections. But is 2.11.0 a minor or a major release of Ansible-base?

jborean93 replied:
Technically from a semver perspective it's a minor release but Ansible doesn't strictly follow semver as we consider the minor part of 2.x being a "major" release. Others might disagree with me here though. I guess it's a question as to whether we see this particular deprecation being done within Ansible itself or only for collections.

Maybe for now we just leave it as is and revisit the message in the future when we don't have a deadline coming up.

felixfontein replied:
With #69680 (adjusted for dates) we would be able to distinguish between ansible-base versions and collection versions. But yeah, let's revisit this somewhen after the freeze :)

jborean93 commented on test/lib/ansible_test/_data/sanity/pylint/plugins/deprecated.py:
```diff
@@ -118,25 +154,41 @@ def visit_call(self, node):
+                if date:
+                    try:
+                        if parse_isodate(date) < datetime.date.today():
+                            self.add_message('ansible-deprecated-date', node=node, args=(date,))
```
I'm not sure if we can actually have this sanity check enabled just by checking the date. I unfortunately don't have a good alternative but this will be problematic for a few reasons.

* The dep by date is meant to enforce that deprecations will have a minimum of 2 years before they are removed, it's not to say that this will be removed at exactly 2 years time
* To remove a feature we would typically need to increase the major version of the collection, doing this at random times just because a dep period has been reached won't paint the best picture of stability of the collection

In short having a sanity test fail when we don't have any immediate plans to actually remove the feature means either we need to ignore the error and defeat the purpose of the check or just live with the error, making CI failures harder to diagnose.

An example scenario of this problem would be the maintenance releases of a certified collection. These collections are currently set to have a rolling release and a 24 month maintenance release. The latter is a bug fix only release with no features being added or removed so it could potentially have a deprecated feature 6 months into it's lifecycle and we need to ignore this sanity check for that branch entirely.

Another scenario is for the rolling release, say we have 2 dep by dates set

1. module x has a `removed_at_date: 2020-06-01`
2. module y has a `removed_at_date: 2020-08-01`

In this scenario we would effectively have to remove the feature at the start of June **and** August to silence the ansible-test sanity checks and in the meantime we either release 2 major bumps in that timeframe or wait the 2 months before doing the next release.

Maybe the best option here is to have this as a manual check, i.e. it's run before creating the release to ensure that all deprecations that can be removed are removed rather than something done in CI.

felixfontein replied:
This test is part of the optional tests, i.e. you explicitly have to run `ansible-test sanity` with `--enable-optional-errors` for it to be a (reported) error. So it will only have an impact if people explicitly request this. (This might change in the future to a non-failing warning, but the result will still be the same: this won't make tests fail from a date on until these tests are explicitly requested. This definitely won't happen for 2.10 though.)

felixfontein replied:
(Optional errors were introduced in #66920, though not used there yet. This would be the first use of that feature.)

jborean93 replied:
Ah that's fantastic, solves this problem quite nicely. Having a quick skim of #66920 it seems like `--enable-optional-errors` is a true/false switch so by setting it you would enable all the optional errors. Is it mean to work in conjuction with the test name like `ansible-test sanity --test ansible-deprecated-date --enable-optional-errors` or is that not possible right now?

felixfontein replied:
`--test` only works with tests like `validate-modules` or `pylint`, not with explicit error codes returned by one of the tests. So you can only enable all of them, or none. @mattclay and me were discussing about that last week, I think the plan is to improve the interface eventually, but not for 2.10. Since it would probably be the only optional error for 2.10 it should be ok for now :)

jborean93 replied:
Fair enough.

jborean93 reviewed (changes requested):
I'm not 100% convinced around the behaviour of the sanity check but I'm happy to discuss it further and try and come to a solution. What are your thoughts on my comments @felixfontein

felixfontein committed:
Update lib/ansible/module_utils/csharp/Ansible.Basic.cs

Co-authored-by: Jordan Borean <jborean93@gmail.com>

felixfontein committed:
Adjust test to latest change.

felixfontein committed:
Prefer date over version if both end up in Display.deprecated().

felixfontein commented:
(Rebased so that the branch includes the changes from the routing PR, and changed behavior of `Display.deprecated()` if both `version` and `date` are specified: prefer `date` in that case. This behavior is now the same as in #69712.)

nitzmahone commented:
rebuild_merge

mattclay commented on test/lib/ansible_test/_data/sanity/pylint/plugins/deprecated.py:
```diff
@@ -64,6 +85,20 @@ def _get_expr_name(node):
+        raise ValueError(msg)
+    # From Python 3.7 in, there is datetime.date.fromisoformat(). For older versions,
+    # we have to do things manually.
+    if not re.match('^[0-9]{4}-[0-9]{2}-[0-9]{2}$', value):
+        raise ValueError(msg)
```
This logic appears to be redundant. The exception handler below should catch the same cases the regex does and return the same exception.

felixfontein replied:
The code below also accepts `2020-1-1`, which is not ISO 8601. It does catch invalid dates such as `2019-02-29` though, which this regex won't catch.

mattclay commented on test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/utils.py:
```diff
@@ -194,3 +197,17 @@ class NoArgsAnsibleModule(AnsibleModule):
+        raise ValueError(msg)
+    # From Python 3.7 in, there is datetime.date.fromisoformat(). For older versions,
+    # we have to do things manually.
+    if not re.match('^[0-9]{4}-[0-9]{2}-[0-9]{2}$', v):
+        raise ValueError(msg)
```
Same as previous comment about redundant logic.

felixfontein commented:
@mattclay @bcoca @sivel @jborean93 @nitzmahone thanks a lot for reviewing and merging! :)
